File: frontend/vicourses_client_web/src/libs/types/lesson.ts
import { VideoFile } from "./common";

export type LessonType = "Video" | "Quiz";

export type Quiz = {
    number: number;
    title: string;
    isMultiChoice: boolean;
    answers: {
        number: number;
        title: string;
        isCorrect: boolean;
        explanation: string | null;
    }[];
};

export type Lesson = {
    id: string;
    courseId: string;
    sectionId: string;
    userId: string;
    title: string;
    order: number;
    type: LessonType;
    createdAt: string;
    updatedAt: string;
    description: string | null;
    video: VideoFile | null;
    quizzesCount: number;
    quizzes: Quiz[];
};

export type CreateLessonRequest = {
    title: string;
    courseId: string;
    sectionId: string;
    type: LessonType;
    description?: string;
}

export type UpdateLessonRequest = {
    title?: string;
    description?: string;
    videoToken?: string;
}

export type AddOrUpdateQuizRequest = {
    title: string;
    answers: {
        title: string;
        isCorrect: boolean;
        explanation?: string;
    }[];
}

File: frontend/vicourses_mobile_app/lib/models/lesson.dart
import 'package:vicourses_mobile_app/models/quiz.dart';
import 'package:vicourses_mobile_app/models/video.dart';

class Lesson {
  final String id;
  final String courseId;
  final String sectionId;
  final String userId;
  final String title;
  final int order;
  final String type;
  final String? description;
  final Video? video;
  final int quizzesCount;
  final List<Quiz> quizzes;

  Lesson({
    required this.id,
    required this.courseId,
    required this.sectionId,
    required this.userId,
    required this.title,
    required this.order,
    required this.type,
    this.description,
    this.video,
    this.quizzesCount = 0,
    required this.quizzes,
  });

  static Lesson fromMap(Map<String, dynamic> data) {
    return Lesson(
      id: data['id'],
      courseId: data['courseId'],
      sectionId: data['sectionId'],
      userId: data['userId'],
      title: data['title'],
      order: data['order'],
      type: data['type'],
      description: data['description'],
      video: data['video'] != null
          ? Video(
              originalFileName: data['video']['originalFileName'],
              duration: data['video']['duration'],
              status: data['video']['status'],
              token: data['video']['token'],
            )
          : null,
      quizzesCount: data['quizzesCount'],
      quizzes: (data['quizzes'] as List).map((e) => Quiz.fromMap(e)).toList(),
    );
  }
}
